### PR TITLE
[mono][infra] Enable runtime tests in fullAOT LLVM mode on linux-x64

### PR DESCRIPTION
**Pull request by kotlarmilos**
## Description

This PR enables jobs to run runtime tests in fullAOT LLVM mode on linux-x64. The jobs utilize the CBL-Mariner docker image, which contains the clang toolchain instead of binutils. Due to OOM and timeout issues, the tests are split into two jobs: intrinsics and other runtime tests.

Failing tests have been disabled and will be triaged in subsequent PRs. The arm64 support will be introduced in subsequent PRs. A list of disabled tests is included in https://github.com/dotnet/runtime/issues/90427.

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

**Comment by kotlarmilos** [comment]
@vargaz The failure has been introduced recently since one of the previous runs successfully compiled tests.
```
2023-10-07T16:45:13.2905201Z   * Assertion at /__w/1/s/src/mono/mono/mini/mini-llvm.c:10032, condition `LLVMTypeOf (lhs) == LLVMTypeOf (rhs)' not met
2023-10-07T16:45:13.2905410Z   
2023-10-07T16:45:13.2905513Z   
2023-10-07T16:45:13.2905660Z   =================================================================
2023-10-07T16:45:13.2905849Z   	Native Crash Reporting
2023-10-07T16:45:13.2906013Z   =================================================================
2023-10-07T16:45:13.2906215Z   Got a SIGABRT while executing native code. This usually indicates
2023-10-07T16:45:13.3107278Z   a fatal error in the mono runtime or one of the native libraries 
2023-10-07T16:45:13.3107628Z   used by your application.
2023-10-07T16:45:13.3107937Z   =================================================================
2023-10-07T16:45:13.3108057Z   
2023-10-07T16:45:13.3108166Z   =================================================================
2023-10-07T16:45:13.3108295Z   	External Debugger Dump:
2023-10-07T16:45:13.3108419Z   =================================================================
2023-10-07T16:45:13.3108526Z   
2023-10-07T16:45:13.3108641Z   =================================================================
2023-10-07T16:45:13.3108756Z   	Native stacktrace:
2023-10-07T16:45:13.3108882Z   =================================================================
2023-10-07T16:45:13.3109351Z   	0x7fe9c1840db1 - Unknown
2023-10-07T16:45:13.3109482Z   	0x7fe9c17e388e - Unknown
2023-10-07T16:45:13.3109614Z   	0x7fe9c1840651 - Unknown
2023-10-07T16:45:13.3109739Z   	0x7fe9c313fcf0 - Unknown
2023-10-07T16:45:13.3110054Z   	0x7fe9c2487acf - Unknown
2023-10-07T16:45:13.3110210Z   	0x7fe9c245aea5 - Unknown
2023-10-07T16:45:13.3110338Z   	0x7fe9c18ff455 - Unknown
2023-10-07T16:45:13.3110470Z   	0x7fe9c190fa46 - Unknown
2023-10-07T16:45:13.3110600Z   	0x7fe9c18ff8bf - Unknown
2023-10-07T16:45:13.3110725Z   	0x7fe9c18ffa25 - Unknown
2023-10-07T16:45:13.3110855Z   	0x7fe9c18ffa67 - Unknown
2023-10-07T16:45:13.3110980Z   	0x7fe9c189a146 - Unknown
2023-10-07T16:45:13.3111111Z   	0x7fe9c1882e13 - Unknown
2023-10-07T16:45:13.3111238Z   	0x7fe9c1880f69 - Unknown
2023-10-07T16:45:13.3111363Z   	0x7fe9c1741b52 - Unknown
2023-10-07T16:45:13.3111493Z   	0x7fe9c17c32a3 - Unknown
2023-10-07T16:45:13.3111617Z   	0x7fe9c17b3914 - Unknown
2023-10-07T16:45:13.3111746Z   	0x7fe9c17af22e - Unknown
2023-10-07T16:45:13.3111880Z   	0x7fe9c17a70cb - Unknown
2023-10-07T16:45:13.3112005Z   	0x7fe9c17a5c00 - Unknown
2023-10-07T16:45:13.3112232Z   	0x7fe9c181f097 - Unknown
2023-10-07T16:45:13.3112358Z   	0x564a9165d6a1 - Unknown
2023-10-07T16:45:13.3112490Z   	0x7fe9c2473d85 - Unknown
2023-10-07T16:45:13.3112737Z   	0x564a9165a7c9 - Unknown
2023-10-07T16:45:13.4584414Z   [New LWP 3397]
2023-10-07T16:45:13.4692486Z EXEC : Dwarf error : Cannot handle DW_FORM_strx1 in DWARF reader [in module /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/corerun] [/__w/1/s/src/mono/msbuild/aot-compile.proj]
2023-10-07T16:45:13.5216383Z   [Thread debugging using libthread_db enabled]
2023-10-07T16:45:13.5216590Z   Using host libthread_db library "/lib64/libthread_db.so.1".
2023-10-07T16:45:13.6867545Z   Missing separate debuginfo for /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.6868382Z   Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/52/46f76caac4d48ccaf38e0ef27d086bfb002231.debug
2023-10-07T16:45:13.7021296Z   0x00007fe9c313f672 in waitpid () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7025005Z     Id   Target Id                                      Frame 
2023-10-07T16:45:13.7025590Z   * 1    Thread 0x7fe9c3773740 (LWP 3396) "corerun"     0x00007fe9c313f672 in waitpid () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7026054Z     2    Thread 0x7fe9c03ff700 (LWP 3397) "SGen worker" 0x00007fe9c313b45c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7078680Z   
2023-10-07T16:45:13.7081067Z   Thread 2 (Thread 0x7fe9c03ff700 (LWP 3397)):
2023-10-07T16:45:13.7081535Z   #0  0x00007fe9c313b45c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7081901Z   #1  0x00007fe9c1a365f3 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7082234Z   #2  0x00007fe9c31351ca in start_thread () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7082532Z   #3  0x00007fe9c2472e73 in clone () from /lib64/libc.so.6
2023-10-07T16:45:13.7084645Z   
2023-10-07T16:45:13.7085296Z   Thread 1 (Thread 0x7fe9c3773740 (LWP 3396)):
2023-10-07T16:45:13.7085534Z   #0  0x00007fe9c313f672 in waitpid () from /lib64/libpthread.so.0
2023-10-07T16:45:13.7085731Z   #1  0x00007fe9c1840ee2 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7086006Z   #2  0x00007fe9c17e388e in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7086380Z   #3  0x00007fe9c1840651 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7088643Z   #4  <signal handler called>
2023-10-07T16:45:13.7088933Z   #5  0x00007fe9c2487acf in raise () from /lib64/libc.so.6
2023-10-07T16:45:13.7091819Z   #6  0x00007fe9c245aea5 in abort () from /lib64/libc.so.6
2023-10-07T16:45:13.7092147Z   #7  0x00007fe9c18ff455 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7092876Z   #8  0x00007fe9c190fa46 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7093347Z   #9  0x00007fe9c18ff8bf in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7093699Z   #10 0x00007fe9c18ffa25 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7097623Z   #11 0x00007fe9c18ffa67 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7098211Z   #12 0x00007fe9c189a146 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7098567Z   #13 0x00007fe9c1882e13 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7098945Z   #14 0x00007fe9c1880f69 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7099281Z   #15 0x00007fe9c1741b52 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7099824Z   #16 0x00007fe9c17c32a3 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7102707Z   #17 0x00007fe9c17b3914 in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7103290Z   #18 0x00007fe9c17af22e in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7103778Z   #19 0x00007fe9c17a70cb in ?? () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7104205Z   #20 0x00007fe9c17a5c00 in mono_main () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7104595Z   #21 0x00007fe9c181f097 in monovm_execute_assembly () from /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/libcoreclr.so
2023-10-07T16:45:13.7105085Z   #22 0x0000564a9165d6a1 in main ()
2023-10-07T16:45:13.7107091Z   [Inferior 1 (process 3396) detached]
2023-10-07T16:45:13.7193105Z   
2023-10-07T16:45:13.7193388Z   =================================================================
2023-10-07T16:45:13.7193568Z   	Basic Fault Address Reporting
2023-10-07T16:45:13.7193733Z   =================================================================
2023-10-07T16:45:13.7193985Z   Memory around native instruction pointer (0x7fe9c2487acf):0x7fe9c2487abf  d2 4c 89 ce bf 02 00 00 00 b8 0e 00 00 00 0f 05  .L..............
2023-10-07T16:45:13.7194232Z   0x7fe9c2487acf  48 8b 8c 24 08 01 00 00 64 48 33 0c 25 28 00 00  H..$....dH3.%(..
2023-10-07T16:45:13.7194423Z   0x7fe9c2487adf  00 44 89 c0 75 19 48 81 c4 10 01 00 00 5b c3 66  .D..u.H......[.f
2023-10-07T16:45:13.7194600Z   0x7fe9c2487aef  90 48 8b 15 71 03 37 00 f7 d8 64 89 02 eb ba e8  .H..q.7...d.....
2023-10-07T16:45:13.8405747Z   /tmp/MSBuildTempcloudtest_azpcontainer/tmpeced0fef58a64d8fa3c9baffaa41f57c.exec.cmd: line 2:  3396 Aborted                 (core dumped) /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/corerun /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll
2023-10-07T16:45:13.8413294Z /__w/1/s/src/mono/msbuild/aot-compile.proj(19,9): error MSB3073: The command "/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/corerun /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll" exited with code 134.
```

**Comment by kotlarmilos** [comment]
This PR is currently blocked due to high Mono AOT memory consumption. Once the https://github.com/dotnet/runtime/issues/95791 is resolved, this PR will be reopened.

**Comment by kotlarmilos** [comment]
@vargaz Please take a look at the fullAOT llvm failure:
```
  Executing the native assembler: "clang" -c -x assembler  -o /tmp/mono_aot_xFbAUF.o /tmp/mono_aot_xFbAUF
  Executing the native linker: "clang" -fuse-ld=lld -shared -o /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.ObjectModel.dll.so.tmp "mono_aot_xO3pCx/temp-llvm.o" /tmp/mono_aot_xFbAUF.o 
  JIT time: 63 ms, Generation time: 1021 ms, Assembly+Link time: 124 ms.
  aot-compile: compiling /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll; MONO_PATH: /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root:/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root
  Mono Ahead of Time compiler - compiling assembly /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll
  AOTID 1B0A2F18-5FC0-C162-5648-F8AAA03474F8
  * Assertion at /__w/1/s/src/mono/mono/mini/mini-llvm.c:8590, condition `elems <= MAX_VECTOR_ELEMS / 2' not met

```

/cc: @fanyang-mono

**Comment by lambdageek** [comment]
There was a failure (on some erarlier build of this PR) where the BStr tests weren't compiling properly.  I added https://github.com/dotnet/runtime/issues/98307

**Comment by fanyang-mono** [comment]
> @vargaz Please take a look at the fullAOT llvm failure:
> 
> ```
>   Executing the native assembler: "clang" -c -x assembler  -o /tmp/mono_aot_xFbAUF.o /tmp/mono_aot_xFbAUF
>   Executing the native linker: "clang" -fuse-ld=lld -shared -o /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.ObjectModel.dll.so.tmp "mono_aot_xO3pCx/temp-llvm.o" /tmp/mono_aot_xFbAUF.o 
>   JIT time: 63 ms, Generation time: 1021 ms, Assembly+Link time: 124 ms.
>   aot-compile: compiling /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll; MONO_PATH: /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root:/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root
>   Mono Ahead of Time compiler - compiling assembly /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll
>   AOTID 1B0A2F18-5FC0-C162-5648-F8AAA03474F8
>   * Assertion at /__w/1/s/src/mono/mono/mini/mini-llvm.c:8590, condition `elems <= MAX_VECTOR_ELEMS / 2' not met
> ```
> 
> /cc: @fanyang-mono

That's fixed by https://github.com/dotnet/runtime/pull/98320

**Comment by fanyang-mono** [comment]
When I run full aot tests locally, I noticed that the new style tests doesn't produce `.sh` files any more. However, the infrastructure relies on it to find the correspoinding `.dll` files to compile. See code here: https://github.com/dotnet/runtime/blob/8cb1f79fd66850e233a4b756aa0b5407e64f1105/src/tests/build.proj#L108

This test is an example of such tests:
src/tests/JIT/HardwareIntrinsics/Arm/AdvSimd/AdvSimd_r.csproj

It would be nice to have this fixed as part of this PR.

**Comment by kotlarmilos** [comment]
> When I run full aot tests locally, I noticed that the new style tests doesn't produce `.sh` files any more. However, the infrastructure relies on it to find the correspoinding `.dll` files to compile. See code here:

I believe the tests for intrinsics are dynamically generated, which could be the reason the `.sh` files are missing. I will check how this is handled in a CoreCLR job and try to include them.

**Comment by fanyang-mono** [comment]
> > When I run full aot tests locally, I noticed that the new style tests doesn't produce `.sh` files any more. However, the infrastructure relies on it to find the correspoinding `.dll` files to compile. See code here:
> 
> I believe the tests for intrinsics are dynamically generated, which could be the reason the `.sh` files are missing. I will check how this is handled in a CoreCLR job and try to include them.

Yes they are dynamically generated. But if you let it fall back to the old style by setting this environment variable `BuildAsStandalone=true`. The corresponding `.sh` file will be generated.

**Comment by fanyang-mono** [comment]
I tried running the failing `Vector128_1_r` locally on my x64 machine. The test passed. @kotlarmilos Were you able to reproduce it locally?

**Comment by vargaz** [comment]
The mono runtime changes look ok to me.

**Comment by kotlarmilos** [comment]
> The mono runtime changes look ok to me.

Thanks. Let's first review and upstream https://github.com/dotnet/runtime/pull/96332.

**Comment by kotlarmilos** [comment]
This PR is ready for review. Failures shouldn't be related.

Runtime tests: https://dev.azure.com/dnceng-public/public/_build/results?buildId=629337&view=logs&jobId=d29f5ac9-2776-59d5-99ad-be6d4bcd416c&j=d29f5ac9-2776-59d5-99ad-be6d4bcd416c
Runtime intrinsics tests: https://dev.azure.com/dnceng-public/public/_build/results?buildId=629337&view=logs&jobId=d29f5ac9-2776-59d5-99ad-be6d4bcd416c&j=8c8e10f8-cac0-5393-e608-4ea6cc126022

**Review by matouskozak — APPROVED** [review]
LGTM! good job

File: eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml [hunk 1]
```diff
@@ -111,43 +111,89 @@ jobs:
...
+    variables:
+      - name: timeoutPerTestInMinutes
+        value: 60
+      - name: timeoutPerTestCollectionInMinutes
+        value: 180
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: AllSubsets_Mono_LLVMFullAot_RuntimeTests_Tests
+      runtimeVariant: llvmfullaot
+      buildArgs: -s mono+libs+clr.hosts+clr.iltools -c Release /p:MonoEnableLLVM=true /p:MonoBundleLLVMOptimizer=true
```
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml:134]: Good point. This command only builds the runtime.

The AOT compiler is built in an optional step which contains required parameters:
https://github.com/dotnet/runtime/blob/e61064949214b9eebdbf9faad58b71afd240ff15/eng/pipelines/common/templates/runtimes/build-runtime-tests-and-send-to-helix.yml#L42-L63

File: eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml [hunk 2]
```diff
@@ -111,43 +111,89 @@ jobs:
...
+    variables:
+      - name: timeoutPerTestInMinutes
+        value: 60
+      - name: timeoutPerTestCollectionInMinutes
+        value: 180
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: AllSubsets_Mono_LLVMFullAot_RuntimeIntrinsicsTests
+      runtimeVariant: llvmfullaot
+      buildArgs: -s mono+libs+clr.hosts+clr.iltools -c Release /p:MonoEnableLLVM=true /p:MonoBundleLLVMOptimizer=true
```
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml:180]: Correct, perhaps we can create a template. @fanyang-mono Should this be in `runtime-llvm.yml`?

**Comment by kotlarmilos** [comment]
These jobs have been moved to `runtime-llvm.yml`. Test run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=630363&view=results

File: eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (matouskozak) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml]: Should we align on the use of `MonoBundleLLVMOptimizer` vs `MonoAOTEnableLLVM` then?

File: eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-other.yml]: Moved to the `runtime-llvm.yml` and fixed `llvmaot` job.

**Comment by kotlarmilos** [comment]
The failures shouldn't be related.

**Pull request merged**